x/streamer/keeper/distribute.go

```diff
@@ -5,110 +5,240 @@ import (
+			// add a new gauge to the cache
+			gaugeCache.Upsert(gauge)
+		}
+
+		rewards, err := k.CalculateGaugeRewards(
+			ctx,
+			v.Stream.EpochCoins,
+			v.Gauge,
+			stream.DistributeTo.TotalWeight,
+		)
 		if err != nil {
-			// Ignore this gauge
+			// we don't want to fail in this case, ignore this gauge
 			k.Logger(ctx).
-				With("streamID", v.Stream.Id, "gaugeID", v.Gauge.GaugeId, "error", err.Error()).
+				With("streamID", stream.Id, "gaugeID", v.Gauge.GaugeId, "error", err.Error()).
 				Error("Failed to distribute to gauge")
-			return pagination.Continue
+			return false, 0 // continue, weight = 0, consider this operation as it is free
 		}
 
-		totalDistributed = totalDistributed.Add(distributed...)
+		// update distributed coins for the stream
+		stream.AddDistributedCoins(rewards)
+		streamCache.Upsert(stream)
 
-		// Update distributed coins for the stream
-		update := streamUpdates[v.Stream.Id]
-		update = update.Add(distributed...)
-		streamUpdates[v.Stream.Id] = update
+		// update distributed coins for the gauge
+		gauge.AddCoins(rewards)
+		gaugeCache.Upsert(gauge)
 
-		return pagination.Continue
+		// get gauge weight and update denomLocksCache under the hood
+		weight = k.getGaugeWeight(ctx, gauge, denomLocksCache)
+
+		distributedCoins = distributedCoins.Add(rewards...)
+
+		return false, weight
 	})
+	return distributedCoins, pointer, iterations
+}
 
-	for i, s := range streams {
-		s.DistributedCoins = s.DistributedCoins.Add(streamUpdates[s.Id]...)
-		streams[i] = s
+// getActiveGaugeByID returns the active gauge with the given ID from the keeper.
+// An error is returned if the gauge does not exist or if it is finished.
+func (k Keeper) getActiveGaugeByID(ctx sdk.Context, gaugeID uint64) (incentivestypes.Gauge, error) {
+	// validate the gauge exists
+	gauge, err := k.ik.GetGaugeByID(ctx, gaugeID)
+	if err != nil {
+		return incentivestypes.Gauge{}, fmt.Errorf("get gauge by id '%d': %w", gaugeID, err)
 	}
+	// validate the gauge is not finished
+	finished := gauge.IsFinishedGauge(ctx.BlockTime())
+	if finished {
+		return incentivestypes.Gauge{}, incentivestypes.UnexpectedFinishedGaugeError{GaugeId: gaugeID}
+	}
+	return *gauge, nil
+}
 
-	return DistributeRewardsResult{
-		NewPointer:       newPointer,
-		FilledStreams:    streams, // Make sure that the returning slice is always sorted
-		DistributedCoins: totalDistributed,
-		Iterations:       iterations,
+// getGaugeWeight calculates the weight of a gauge based on its type.
+// If the gauge is an asset gauge, the weight is equal to the number of associated lockups. Use cache to get this figure.
+// If the gauge is a rollapp gauge, the weight is always 1. If the gauge has an unknown type, the weight is 0 as we
+// assume that the gauge is always validated at this step and has a known type. The method also fills the cache under
+// the hood.
+func (k Keeper) getGaugeWeight(ctx sdk.Context, gauge incentivestypes.Gauge, cache incentivestypes.DenomLocksCache) uint64 {
```

Comment:
the weight terminology is confusing because you have weight for people allocating different weights to different guages, and then the weight concept of iteration



Possible to use another term?